### PR TITLE
Introduce and implement image GC policy

### PR DESCRIPTION
**Pull request by vmarmol**

### CODE REVIEW

File: pkg/kubelet/image_manager_test.go [hunk 1]
```diff
@@ -21,17 +21,39 @@ import (
...
 
-func newRealImageManager(dockerClient dockertools.DockerInterface) *realImageManager {
-	return newImageManager(dockerClient).(*realImageManager)
+func newRealImageManager(t *testing.T, policy ImageGCPolicy) (*realImageManager, *dockertools.FakeDockerClient, *cadvisor.Mock) {
+	fakeDocker := &dockertools.FakeDockerClient{
+		RemovedImages: util.NewStringSet(),
+	}
+	mockCadvisor := new(cadvisor.Mock)
+	im, err := newImageManager(fakeDocker, mockCadvisor, policy)
+	require.Nil(t, err)
```
Comment (vishh) [inline pkg/kubelet/image_manager_test.go]: nit: Use `require.NoError` instead

File: pkg/kubelet/cadvisor/cadvisor_fake.go [hunk 2]
```diff
@@ -18,6 +18,7 @@ package cadvisor
 
 import (
 	cadvisorApi "github.com/google/cadvisor/info/v1"
+	cadvisorApi2 "github.com/google/cadvisor/info/v2"
```
Comment (vishh) [inline pkg/kubelet/cadvisor/cadvisor_fake.go]: nit: `cadvisorApiv2` here and elsewhere.

**Comment by vishh** [comment]
LGTM. This feature will be super useful. Thanks @vmarmol!

File: pkg/kubelet/image_manager_test.go (hunk 1 above)
Comment (vmarmol) [inline pkg/kubelet/image_manager_test.go]: Done.

File: pkg/kubelet/cadvisor/cadvisor_fake.go (hunk 2 above)
Comment (vmarmol) [inline pkg/kubelet/cadvisor/cadvisor_fake.go]: Done.

File: pkg/kubelet/image_manager.go [hunk 3]
```diff
@@ -144,7 +171,39 @@ func (self *realImageManager) detectImages(detected time.Time) error {
...
+	// If over the max threshold, free enough to place us at the lower threshold.
+	if usagePercent >= self.policy.HighThresholdPercent {
+		amountToFree := usage - (int64(self.policy.LowThresholdPercent) * capacity / 100)
+		freed, err := self.freeSpace(amountToFree)
+		if err != nil {
+			return err
+		}
+
+		if freed < amountToFree {
+			return fmt.Errorf("failed to garbage collect required amount of images. Wanted to free %d, but freed %d", amountToFree, freed)
```
Comment (brendandburns) [inline pkg/kubelet/image_manager.go]: Can you at least TODO making this a true error that you can introspect on?

File: pkg/kubelet/image_manager.go [hunk 4]
```diff
@@ -29,31 +31,40 @@ import (
...
-	FreeSpace(bytesToFree int64) (int64, error)
+	// Applies the garbage collection policy. Errors include being unable to free
+	// enough space as per the garbage collection policy.
+	GarbageCollect() error
 
 	// TODO(vmarmol): Have this subsume pulls as well.
 }
 
+// A policy for garbage collecting images. Policy defines an allowed band in
+// which garbage collection can be run.
```
Comment (brendandburns) [inline pkg/kubelet/image_manager.go]: "will be run." ?

**Comment by brendandburns** [comment]
This LGTM, really minor comments.

File: pkg/kubelet/image_manager.go (hunk 3 above)
Comment (vmarmol) [inline pkg/kubelet/image_manager.go]: Ah, good point we should surface an event. Will add the TODO for now.

File: pkg/kubelet/image_manager.go (hunk 4 above)
Comment (vmarmol) [inline pkg/kubelet/image_manager.go]: Done.

**Comment by dchen1107** [comment]
LGTM. But here are some concerns I have (just for discussion, no need to change for now):

GC images really depends on GC container. If there is a dead container still pointing to the image, Kubelet cannot remove that image even we hit the HighThresholdPercent. We introduced per node cap for dead containers, which mitigate the possibilities running into such issue, but still have chance. In a long run, I think we should combine GC image and container together. When HighThresholdPercent, we start to remove old dead containers even not hit the container cap, then image. Even more, I think we should expand this to disk management eventually. 

Will merge it on green.

**Comment by vmarmol** [comment]
+1 @dchen1107, I also see these merging in the long run.

**Pull request merged**